### PR TITLE
Categorize member Challenge

### LINKED ISSUE #28: Categorize Member Challenge
## Description

Categorize Member Challenge
---

## Behavior

Categorizes an individual into a membership program based on their age and handicap

---

#### Input:
categorize_member(17, 2)
#### Output:
"'Junior'"

#### Input:
categorize_member(55, 8)
#### Output:
"Senior"

#### Input:
categorize_member(25, 5)
#### Output:
"Intermediate"

#### Input:
categorize_member(45, 9)
#### Output:
"Open"

### PR DESCRIPTION
**Pull request by KarimMakki**
---
name: Categorize member Challenge
about: A template PR for code review with a checklist
---

## Behavior
Categorizes an individual into a membership program based on their age and handicap

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/categorize_member.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    on age and skill level
+
+Created on: 2025-01-07
+@author: KarimMakki
+"""
+
+
+def categorize_member(age: int, handicap: int) -> str:
+    """Categorizes an individual into a membership program based on their age and handicap.
+
```
Comment (nico1103) [inline solutions/categorize_member.py:18]: I think adding the explanation of the behavior, how it is categorized according to age and skill, will make it better.

File: solutions/categorize_member.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    Parameters:
+        age (int): The individual's age. Must be non-negative.
+        handicap (int): The individual's skill level. Must be between 0 and 10.
```
Comment (nico1103) [inline solutions/categorize_member.py]: what about 0 value for age? I dont know what the categorization is for but i am guessing it does not make much sense for the age 0 to be categorized as juniors especially it is also taking skill into account and considering toddlers would be less appropriate in my opinion. Considering this, from my pov, the minimum age should be 10.

File: solutions/categorize_member.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    Returns:
+        str: The membership category ("Junior", "Senior", "Intermediate", or "Open")
+
+    Raises:
+        AssertionError:
+        -  If age is not an integer
+        -  If handicap is not an integer
```
Comment (nico1103) [inline solutions/categorize_member.py:30]: As far as im concerned, AssertionErrors are used by programmers to catch logical bugs, while ValueErrors happen due to user input. Therefore, if you are considering these bugs as user inputs, these should be categorized as value errors.

File: solutions/categorize_member.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return "Junior"
+    # Age is 55 or older AND handicap in range 8-10
+    elif age >= 55 and 8 <= handicap <= 10:
+        return "Senior"
+    # Age is 18 or older AND handicap in range 4-7
+    elif age >= 18 and 4 <= handicap <= 7:
+        return "Intermediate"
+    # Any other case
+    else:
+        return "Open"
```
Comment (nico1103) [inline solutions/categorize_member.py]: What happens when one input is in one range, but, another input is in the range of another category? for e.g. if age is 17 and handicap is 4?

**Review by nico1103 — CHANGES_REQUESTED** [review]

File: solutions/categorize_member.py (hunk 1 above)
Comment (KarimMakki) [reply solutions/categorize_member.py:18]: I believe that's more of a strategy than the behaviour in which we should include in the comments between the lines of code rather in the docstrings, that's why I didn't do it in the docstrings but in comments.

File: solutions/categorize_member.py (hunk 2 above)
Comment (KarimMakki) [reply solutions/categorize_member.py]: I agree with you! modified it so that the minimum age must not be less than 10, this should make more sense now.

File: solutions/categorize_member.py (hunk 3 above)
Comment (KarimMakki) [reply solutions/categorize_member.py:30]: I had chosen AssertionError to indicate a violation of an assumption within the function and that what's mentioned in the code review checklist in the implementation section to include defensive assertions.

File: solutions/categorize_member.py (hunk 4 above)
Comment (KarimMakki) [reply solutions/categorize_member.py]: For that case, it should return `"Open"` as it does not meet the condition for the other cases.

Comment (nico1103) [reply solutions/categorize_member.py]: I think "open" sounds a bit ambiguous in my opinion. Would you like to add a bit more explanation onto that? (After all, age is not directly proportional to the skill, so, there would be many cases where the skill level is high while the age is not within a senior range.)

File: solutions/categorize_member.py (hunk 1 above)
Comment (nico1103) [reply solutions/categorize_member.py:18]: Alright, I think you are probably right that it falls under the strategy.

File: solutions/categorize_member.py (hunk 3 above)
Comment (nico1103) [reply solutions/categorize_member.py:30]: I think we have some differences in our understanding but I dont think it is much of an issue. What's more important is you made sure there are errors raised as defensive cases. So, I think **it is alright to proceed**.

**Review by nico1103 — CHANGES_REQUESTED** [review]
I have made one request and passed the others.

File: solutions/categorize_member.py (hunk 4 above)
Comment (KarimMakki) [reply solutions/categorize_member.py]: Thanks for the heads up @nico1103! I added more categories now to handle any case we discussed and tested them

File: solutions/categorize_member.py [hunk 5]
```diff
@@ -35,12 +36,19 @@ def categorize_member(age: int, handicap: int) -> str:
...
         'Intermediate'
+        >>> categorize_member(30, 2)
+        'Low Skill Intermediate'
         >>> categorize_member(45, 9)
-        'Open'
+        'High Skill Intermediate'
+        >>> categorize_member(55, 8)
+        'Senior'
+        >>> categorize_member(60, 3)
+        'Low Skill Senior'
```
Comment (nico1103) [inline solutions/categorize_member.py]: you have got too many tests in the docstring. you might want to narrow it down to max 4.

**Review by nico1103 — CHANGES_REQUESTED** [review]
Only one remark to comply.

File: solutions/categorize_member.py (hunk 5 above)
Comment (KarimMakki) [reply solutions/categorize_member.py:50]: It's actually needed in my case because I want to have a doctest for each category that's why, as long as they are passing then it should be okay.

Comment (KarimMakki) [reply solutions/categorize_member.py]: Sure, I narrowed them down to 4 now!

**Review by nico1103 — APPROVED** [review]
